frontend/storm_wiki/util/search.py

```diff
@@ -0,0 +1,230 @@
+import os
+import re
+import json
+import requests
+import xml.etree.ElementTree as ET
+from urllib.parse import urlparse
+from typing import Union, List, Dict, Any
+import dspy
+import streamlit as st
+from langchain_community.utilities.duckduckgo_search import DuckDuckGoSearchAPIWrapper
+
+from pages_util.Settings import load_search_options
+
+import logging
+
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+)
+logger = logging.getLogger(__name__)
+
+
+class CombinedSearchAPI(dspy.Retrieve):
+    def __init__(self, max_results=20):
+        super().__init__()
+        self.max_results = max_results
+        self.search_options = load_search_options()
+        self.primary_engine = self.search_options["primary_engine"]
+        self.fallback_engine = self.search_options["fallback_engine"]
+        self.ddg_search = DuckDuckGoSearchAPIWrapper()
+        self.searxng_base_url = (
+            self.search_options.get("engine_settings", {})
+            .get("searxng", {})
+            .get("base_url", "http://localhost:8080")
+        )
+        self.search_engines = self._initialize_search_engines()
+        self._initialize_domain_restrictions()
+
+    def _initialize_search_engines(self):
```

Comment:
`yourdm` is not registered here. This is a bit confusing as I run into the issue in first run because I add `YDC_API_KEY` and select the search engine as `yourdm`.

Reply:
sure. same with bing search.